File: hackerrank/sherlockAndAnagrams.js
// https://www.hackerrank.com/challenges/sherlock-and-anagrams/problem

function sherlockAndAnagrams(s) {
  const alphabet = 'abcdefghijklmnopqrstuvwxyz'
  const signatures = {}
  let res = 0

  for (const start in s) {
    for (let finish = start; finish < s.length; finish++) {
      let signature = (new Array(alphabet.length)).fill(0)
      for (const letter of s.slice(start, +finish + 1)) {
        signature[letter.charCodeAt() - alphabet[0].charCodeAt()] += 1
      }
      signature = signature.join('')
      if (signatures[signature] === undefined)
        signatures[signature] = 1
      else
        signatures[signature] += 1
      console.log(s.slice(start, +finish + 1))
      console.log(signatures)
    }
  }

  for (const count of Object.values(signatures)) {
    res += count * (count - 1) / 2
  }

  return res
}

console.log(sherlockAndAnagrams('aa'))         //=> 1
console.log(sherlockAndAnagrams('abba'))       //=> 4
// console.log(sherlockAndAnagrams('abcd'))       //=> 0
// console.log(sherlockAndAnagrams('ifailuhkqq')) //=> 3
// console.log(sherlockAndAnagrams('kkkk'))       //=> 10
// console.log(sherlockAndAnagrams('cdcd'))       //=> 5

File: hackerrank/sherlockAndAnagrams.py
# https://www.hackerrank.com/challenges/sherlock-and-anagrams/problem

import string

def sherlockAndAnagrams(s):
    ALPHABET = string.ascii_lowercase
    signatures = {}

    # iterate over all substrings of s
    for start in range(len(s)):
        for finish in range(start, len(s)):
            # initialize substring signature
            signature = [0 for _ in ALPHABET]
            for letter in s[start:finish+1]:
                signature[ord(letter)-ord(ALPHABET[0])] += 1
            # tuples are hashable in contrast to lists
            signature = tuple(signature)
            signatures[signature] = signatures.get(signature, 0) + 1
    
    res = 0
    for count in signatures.values():
        res += count*(count-1)/2
    return res

# print(sherlockAndAnagrams('aa'))           #=> 1
print(sherlockAndAnagrams('abba'))       #=> 4
# print(sherlockAndAnagrams('abcd'))       #=> 0
# print(sherlockAndAnagrams('ifailuhkqq')) #=> 3
# print(sherlockAndAnagrams('kkkk'))       #=> 10
# print(sherlockAndAnagrams('cdcd'))       #=> 5